### PR TITLE
KILL-THREAD-SLOWLY

### PR DESCRIPTION
**Pull request by appleby**
Add a `(SLEEP 1)` after every `BT:DESTROY-THREAD` call.

This is a work-around that enables switching over to EC2 runners for the gitlab CI. Something about the EC2 environment was tickling a race in our combination of using `BT:DESTROY-THREAD` + our reliance on `PZMQ`'s implicit global `*DEFAULT-CONTEXT*`.

A proper fix for this likely entails some combination of

1) Explicitly creating our own `PZMQ:CONTEXT` in `RPCQ:START-SERVER` and passing that down to worker threads.

2) Adding a graceful shutdown option to the `RPCQ` server to avoid the need to `BT:DESTROY-THREAD` (see #75).

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ variables:
 test-python:
   stage: test
   tags:
-    - github
+    - ec2-github
```
Comment (appleby) [inline .gitlab-ci.yml]: I don't know if this is intended to get merged, or just meant for testing EC2 runners. If the latter, I can drop it.

File: src-tests/test-rpc.lisp [hunk 2]
```diff
@@ -4,6 +4,20 @@
 
 (in-package #:rpcq-tests)
 
+;; TODO FIXME XXX The below relies on some funky/hacky/unsightly code
```
Comment (jmbr) [inline src-tests/test-rpc.lisp]: Perhaps it could be desirable to have this also as an issue on Github.

File: src-tests/test-rpc.lisp [hunk 3]
```diff
@@ -4,6 +4,20 @@
...
+;; https://github.com/rigetti/rpcq/issues/75
+
+(defun kill-thread-slowly (thread)
+  #+ccl
+  (loop :while (bt:thread-alive-p thread)
+        :do (sleep 1) (bt:destroy-thread thread))
+  #-ccl
+  (when (bt:thread-alive-p thread)
+    (bt:destroy-thread thread))
+  (sleep 1))
```
Comment (jmbr) [inline src-tests/test-rpc.lisp]: I'm curious about why do you need to wait, then kill in CCL whereas you kill, then wait on other compilers.

File: src-tests/test-rpc.lisp [hunk 4]
```diff
@@ -52,11 +61,7 @@
              (let ((server-response (rpcq:rpc-call client "test-method")))
                (is (string= *expected-response* server-response))))
         ;; kill the server thread
-        #+ccl
-        (loop :while (bt:thread-alive-p server-thread)
-              :do (sleep 1) (bt:destroy-thread server-thread))
-        #-ccl
-        (bt:destroy-thread server-thread)))))
+        (kill-thread-slowly server-thread)))))
```
Comment (jmbr) [inline src-tests/test-rpc.lisp]: Killing them softly.

**Comment by appleby** [comment]
I'm planning to follow this up "real soon" with a "real fix", but wanted to unblock the wheels of progress on ec2 enabling.

Of course, by the law of Quick-and-Dirty Temporary Fixes, this hack will probably outlive us all.

File: src-tests/test-rpc.lisp (hunk 3 above)
Comment (appleby) [reply src-tests/test-rpc.lisp]: The CCL sleep was pre-existing. I considered flipping the order, but decided to let a sleeping dog lie 🥁 .

Possibly the single sleep would be enough to also fix CCL shutdown, but I didn't test it.

File: src-tests/test-rpc.lisp (hunk 2 above)
Comment (appleby) [reply src-tests/test-rpc.lisp]: I consider fixing this to be part of #75, but I've added a note there now to make sure this doesn't fall through the cracks.

https://github.com/rigetti/rpcq/issues/75#issuecomment-574375441

File: src-tests/test-rpc.lisp (hunk 4 above)
Comment (appleby) [reply src-tests/test-rpc.lisp]: I almost named this function `KILLING-THREAD-SOFTLY` :-)

**Pull request merged**